swarm/network/discovery_test.go

```diff
@@ -58,3 +68,164 @@ func TestDiscovery(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+const (
+	maxPO     = 10
+	maxPeerPO = 8
+)
+
+// TestInitialPeersMsg tests if peersMsg response to incoming subPeersMsg is correct
+func TestSubPeersMsg(t *testing.T) {
+	for po := 0; po < maxPO; po++ {
+		for depth := 0; depth < maxPO; depth++ {
+			t.Run(fmt.Sprintf("PO=%d,advetised depth=%d", po, depth), func(t *testing.T) {
+				testSubPeersMsg(t, po, depth)
+			})
+		}
+	}
+}
+
+// testSubPeersMsg tests that the correct set of peer info is sent
+// to another peer after receiving their subPeersMsg request
+func testSubPeersMsg(t *testing.T, peerPO, peerDepth int) {
+	// generate random pivot address
+	prvkey, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	pivotAddr := pot.NewAddressFromBytes(PrivateKeyToBzzKey(prvkey))
+	// generate control peers address at peerPO wrt pivot
+	peerAddr := pot.RandomAddressAt(pivotAddr, peerPO)
+	// construct kademlia and hive
+	to := NewKademlia(pivotAddr[:], NewKadParams())
+	hive := NewHive(NewHiveParams(), to, nil)
+
+	// expected addrs in peersMsg response
+	var expBzzAddrs []*BzzAddr
+	addrAt := func(a pot.Address, po int) []byte {
+		b := pot.RandomAddressAt(a, po)
+		return b[:]
+	}
+	connect := func(base pot.Address, po int) *BzzAddr {
+		on := addrAt(base, po)
+		peer := newDiscPeer(on)
+		hive.On(peer)
+		return peer.BzzAddr
+	}
+	register := func(base pot.Address, po int) {
+		hive.Register(&BzzAddr{OAddr: addrAt(base, po)})
+	}
+
+	for po := maxPeerPO; po >= 0; po-- {
+		// create a fake connected peer at po from peerAddr
+		on := connect(peerAddr, po)
+		// create a fake registered address at po from peerAddr
+		register(peerAddr, po)
+		// we collect expected peer addresses only up till peerPO
+		if po < peerDepth {
+			continue
+		}
+		expBzzAddrs = append(expBzzAddrs, on)
+	}
+
+	// create a special bzzBaseTester in which we can associate `enode.ID` to the `bzzAddr` we created above
+	s, _, err := newBzzBaseTesterWithAddrs(t, prvkey, [][]byte{peerAddr[:]}, DiscoverySpec, hive.Run)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// peerID to use in the protocol tester testExchange expect/trigger
+	peerID := s.Nodes[0].ID()
+
+	// now we need to wait until the tester's control peer appears in the hive
+	// so the protocol started
+	ticker := time.NewTicker(10 * time.Millisecond)
+	attempts := 100
+	for range ticker.C {
```

Comment:
What do you think about:

```
found := false
for attempts := 0; attempts  < 100; attempts++ {
  if _, found = hive.peers[peerID]; found {
   break
  }
  time.Sleep(10 * time.Millisecond)
}

if !found {
 t.Fatal(" .... ")
}
```


Would that be simpler? As it:
- only uses the most simple timer possible
- what we do with attempts is obvious from just the loop header
- pre-declaring `found:=false` makes it clear we are doing a pessimistic search

Reply:
eliminated the need